### PR TITLE
Merge exp into flaml

### PR DESCRIPTION
**Pull request by liususan091219**

### CODE REVIEW

File: flaml/nlp/result_analysis/azure_utils.py [hunk 1]
```diff
@@ -711,9 +844,102 @@ def get_config_and_score_from_matched_blob_list(self,
...
             each_config_and_score_list = ConfigScoreList(
                 jobid_config=each_jobconfig,
                 blob_file=each_blob,
-                config_score_list=[ConfigScore(**each_dict) for each_dict in data_json['val_log']])
+                config_score_list=[ConfigScore(**each_dict) for each_dict in data_json['val_log']],
+                test_metric=test_metric)
             matched_config_score_lists.append(each_config_and_score_list)
         return matched_config_score_lists
+
+    def get_plot_data(self, sorted_ts, this_ts2score):
```
Comment (sonichi) [inline flaml/nlp/result_analysis/azure_utils.py:859]: test coverage for all new code

File: test/hf/run_analysis.py [hunk 2]
```diff
@@ -0,0 +1,760 @@
+'''Require: pip install torch transformers datasets wandb flaml[blendsearch,ray]
```
Comment (sonichi) [inline test/hf/run_analysis.py:1]:
```suggestion
'''Require: pip install flaml[nlp,ray]
```

File: test/hf/run_analysis.py [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+'''
```
Comment (sonichi) [inline test/hf/run_analysis.py:2]: I didn't review this file. Is it finalized or in progress? If in progress, you can exclude it from this PR.

File: test/hf/run_autohf.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+    # _exhaustive_sweep(console_args, jobid_config, autohf, wandb_utils)
+
+    # evaluate_configs(autohf, console_args)
+
+    if console_args.algo_mode.endswith("cv"):
+        evaluate_configs_cv(autohf, console_args, 10, wandb_utils)
+    elif console_args.algo_mode.endswith("hpo"):
+        _test_hpo(console_args, jobid_config, autohf, wandb_utils)
+    else:
+        modelcard_exp(console_args)
```
Comment (sonichi) [inline test/hf/run_autohf.py:507]: What's the purpose of run_autohf.py and run_analysis.py respectively? Are they both ready for this PR?

**Review by sonichi — CHANGES_REQUESTED** [review]
This is round 1 of review.
Please also follow the pre-commit instructions as my comments of #208

File: flaml/nlp/result_analysis/azure_utils.py [hunk 5]
```diff
@@ -153,10 +159,26 @@ def is_match(self, partial_jobid):
         """
         is_not_match = False
         for key, val in partial_jobid.__dict__.items():
+            if key == "pre": continue
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/nlp/result_analysis/azure_utils.py]: The definition can be found in the documentation under class JobID

File: flaml/nlp/result_analysis/azure_utils.py [hunk 6]
```diff
@@ -153,10 +159,26 @@ def is_match(self, partial_jobid):
...
+                val = set([str(x) for x in val])
+                this_value = getattr(self, key)
+                if isinstance(this_value, int) or isinstance(this_value, str):
+                    this_value = set([str(this_value)])
+                is_subset = len(this_value.difference(val)) == 0
+                if is_subset is False:
+                    is_not_match = True
+            else:
+                each_val = getattr(self, key)
+                if key == "dat":
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/nlp/result_analysis/azure_utils.py:187]: The definition can be found in the documentation under class JobID

File: flaml/nlp/result_analysis/azure_utils.py [hunk 7]
```diff
@@ -153,10 +159,26 @@ def is_match(self, partial_jobid):
...
+                    each_val = [x.replace("_", "-") for x in each_val]
+                    val = [x.replace("_", "-") for x in val]
+                if key == "pre_full":
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/nlp/result_analysis/azure_utils.py:190]: The definition can be found in the documentation under class JobID

File: test/hf/run_analysis.py (hunk 3 above)
Comment (liususan091219) [reply test/hf/run_analysis.py]: It's in progress, I'll exclude it before you are ready to approve

File: test/hf/run_autohf.py (hunk 4 above)
Comment (liususan091219) [reply test/hf/run_autohf.py:644]: They'll both be excluded. They are here for debugging.

File: test/hf/run_analysis.py (hunk 2 above)
Comment (liususan091219) [reply test/hf/run_analysis.py]: addressed

File: flaml/nlp/result_analysis/azure_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/nlp/result_analysis/azure_utils.py]: It's too far away. Please do one of the following: rename 'pre' to a more informative name; or add comments.

File: flaml/nlp/result_analysis/azure_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/nlp/result_analysis/azure_utils.py:187]: It's too far away. Please do one of the following: rename 'pre' to a more informative name; or add comments.

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 
 repos:
   - repo: https://github.com/psf/black
-    rev: stable
+    rev: 21.9b0
```
Comment (sonichi) [inline .pre-commit-config.yaml:4]: Please don't change this file.

File: benchmarkeval/amlk8s_hyperpartisan.yml [hunk 9]
```diff
@@ -0,0 +1,65 @@
+description: finetuning test on AMLK8s
```
Comment (sonichi) [inline benchmarkeval/amlk8s_hyperpartisan.yml:1]: Remove this file.

**Review by sonichi — CHANGES_REQUESTED** [review]
More changed requested.

**Review by sonichi — COMMENTED** [review]
partial review

File: flaml/nlp/autotransformers.py [hunk 10]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+                        )
+
+                    for each_hp in list(each_init_config.keys()):
+                        assert (
+                            isinstance(each_init_config[each_hp], int)
+                            or isinstance(each_init_config[each_hp], float)
+                            or isinstance(each_init_config[each_hp], str)
+                            or isinstance(each_init_config[each_hp], bool)
+                        ), " points_to_evaluate must be a scalar"
+
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:674]:
```suggestion
                            print(each_hp, "is not in the search space, deleting from init config")
                            continue
                        hp_value = each_init_config[each_hp]
                        domain = self._search_space_hpo[each_hp]
```

File: flaml/nlp/autotransformers.py [hunk 11]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+                                self._search_space_hpo[each_hp],
+                                ray.tune.sample.Categorical,
+                            )
+                            or isinstance(
+                                self._search_space_hpo[each_hp], ray.tune.sample.Float
+                            )
+                            or isinstance(
+                                self._search_space_hpo[each_hp], ray.tune.sample.Integer
+                            )
+                        ), "Every hp space must either be categorical, integer or float"
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:686]: This check shouldn't be here. Also, the domain can be a constant.
```suggestion
```

File: flaml/nlp/autotransformers.py [hunk 12]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+
+                        if isinstance(
+                            self._search_space_hpo[each_hp], ray.tune.sample.Categorical
+                        ):
+                            assert (
+                                each_init_config[each_hp]
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:692]:
```suggestion
                                hp_value
```

File: flaml/nlp/autotransformers.py [hunk 13]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
-                        assert isinstance(self._search_space_hpo[each_hp], ray.tune.sample.Categorical) or \
-                               isinstance(self._search_space_hpo[each_hp], ray.tune.sample.Float) or \
-                               isinstance(self._search_space_hpo[each_hp], ray.tune.sample.Integer), \
-                               "Every hp space must either be categorical, integer or float"
-
-                        if isinstance(self._search_space_hpo[each_hp], ray.tune.sample.Categorical):
-                            assert each_init_config[each_hp] in self._search_space_hpo[each_hp].categories, \
-                                "points_to_evaluate {each_hp} value must be within the search space"
+                    for each_hp in list(each_init_config.keys()):
+                        if each_hp not in self._search_space_hpo.keys():
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:659]:
```suggestion
                        if each_hp not in self._search_space_hpo:
```

File: flaml/nlp/autotransformers.py [hunk 14]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+                            or isinstance(
+                                self._search_space_hpo[each_hp], ray.tune.sample.Float
+                            )
+                            or isinstance(
+                                self._search_space_hpo[each_hp], ray.tune.sample.Integer
+                            )
+                        ), "Every hp space must either be categorical, integer or float"
+
+                        if isinstance(
+                            self._search_space_hpo[each_hp], ray.tune.sample.Categorical
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:689]:
```suggestion
                            domain, ray.tune.sample.Categorical
```

File: flaml/nlp/autotransformers.py [hunk 15]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+                        ):
+                            assert (
+                                each_init_config[each_hp]
+                                in self._search_space_hpo[each_hp].categories
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:693]:
```suggestion
                                in domin.categories
```

File: flaml/nlp/autotransformers.py [hunk 16]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
+                            ), "points_to_evaluate {each_hp} value must be within the search space"
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:694]:
```suggestion
                            ), f"points_to_evaluate {each_hp} value must be within the search space"
```

File: flaml/nlp/autotransformers.py [hunk 17]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
                         else:
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:695]:
```suggestion
                        elif isinstance(domain, ray.tune.sample.Float) or isinstance(domain, ray.tune.sample.Integer):
```

File: flaml/nlp/autotransformers.py [hunk 18]
```diff
@@ -428,40 +651,64 @@ def model_init():
...
-                         search_algo_args_mode,
-                         time_budget,
-                         metric_name,
-                         metric_mode_name,
-                         **custom_hpo_args):
+                            assert (
+                                self._search_space_hpo[each_hp].lower
+                                <= each_init_config[each_hp]
+                                <= self._search_space_hpo[each_hp].upper
+                            ), "points_to_evaluate {each_hp} value must be within the search space"
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:700]:
```suggestion
                                domain.lower <= hp_value <= domain.upper
                            ), f"points_to_evaluate {each_hp} value must be within the search space"
```

**Review by sonichi — COMMENTED** [review]
review in progress

File: flaml/nlp/dataset/dataprocess_auto.py [hunk 19]
```diff
@@ -115,38 +128,53 @@ def tokenize_superglue_wic(this_example,
...
         sentence2_key = None
 
     data_pair = (
-        (this_example[sentence1_key],) if sentence2_key is None else (
-            this_example[sentence1_key], this_example[sentence2_key])
+        (this_example[sentence1_key],)
+        if sentence2_key is None
+        else (this_example[sentence1_key], this_example[sentence2_key])
     )
     assert "max_seq_length" in kwargs, "max_seq_length must be provided for glue"
```
Comment (sonichi) [inline flaml/nlp/dataset/dataprocess_auto.py:159]: move to the beginning of the function

File: flaml/nlp/dataset/dataprocess_auto.py [hunk 20]
```diff
@@ -115,38 +128,53 @@ def tokenize_superglue_wic(this_example,
...
         sentence1_key, sentence2_key = sentence_keys[0], sentence_keys[1]
     else:
-        # TODO coverage
         sentence1_key = sentence_keys[0]
         sentence2_key = None
 
     data_pair = (
-        (this_example[sentence1_key],) if sentence2_key is None else (
-            this_example[sentence1_key], this_example[sentence2_key])
+        (this_example[sentence1_key],)
```
Comment (sonichi) [inline flaml/nlp/dataset/dataprocess_auto.py:155]:
```suggestion
        (this_example[sentence1_key], )
```

File: flaml/nlp/dataset/dataprocess_auto.py [hunk 21]
```diff
@@ -178,12 +208,14 @@ def __init__(self):
...
     @classmethod
-    def from_model_and_dataset_name(cls,
-                                    data_raw,
-                                    model_checkpoint_path,
-                                    dataset_name_list: list = None,
-                                    subdataset_name=None,
-                                    **kwargs):
+    def from_model_and_dataset_name(
+        cls,
+        subfold_dataset,
```
Comment (sonichi) [inline flaml/nlp/dataset/dataprocess_auto.py:213]: update docstr

File: flaml/nlp/dataset/dataprocess_auto.py [hunk 22]
```diff
@@ -212,19 +244,29 @@ def from_model_and_dataset_name(cls,
...
+                batched=False,
+            )
         raise ValueError(
             "Unrecognized method {},{} for this kind of AutoGridSearchSpace: {}.\n"
             "Method name should be one of {}.".format(
-                dataset_name, subdataset_name, cls.__name__, ", ".join(c[0] for c in TOKENIZER_MAPPING.keys())
+                dataset_name,
+                subdataset_name,
+                cls.__name__,
+                ", ".join(c[0] for c in TOKENIZER_MAPPING.keys()),
```
Comment (sonichi) [inline flaml/nlp/dataset/dataprocess_auto.py:270]:
```suggestion
                ", ".join(c[0] for c in TOKENIZER_MAPPING),
```

**Review by sonichi — COMMENTED** [review]
two files reviewed.

File: flaml/nlp/autotransformers.py [hunk 23]
```diff
@@ -82,16 +82,22 @@ def _set_search_space(self, **custom_hpo_args):
...
+            if "grid_search_space" in custom_hpo_args.keys():
+                grid_space_model_type = custom_hpo_args["grid_search_space"]
+            elif self.jobid_config.pre in GRID_SEARCH_SPACE_MAPPING.keys():
+                grid_space_model_type = self.jobid_config.pre
             else:
+                # grid mode is not designed for user to use. Therefore we only allow
+                # grid search in the following two cases: (1) if
+                # user has specified a model, and the model contains a grid space in FLAML
+                # or (2) the user has specified which model's grid space to use
                 raise NotImplementedError("The grid space is not implemented in FLAML")
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:94]:
```suggestion
                raise NotImplementedError(f"The grid space for {self.jobid_config.pre} is not implemented in FLAML.")
```

**Review by sonichi — COMMENTED** [review]
autotransformers

File: flaml/nlp/README.md [hunk 24]
```diff
@@ -8,20 +8,39 @@ An example of using AutoTransformers:
...
+
+autohf_settings = {
+    "resources_per_trial": {"cpu": 1, "gpu": 1},
+    "num_samples": -1,
+    "time_budget": 300,
+}
+
 validation_metric, analysis = autohf.fit(**autohf_settings)
+if validation_metric is not None:
+    predictions, test_metric = autohf.predict()
```
Comment (sonichi) [inline flaml/nlp/README.md:42]: Don't change user-facing API unless necessary.

File: flaml/nlp/autotransformers.py [hunk 25]
```diff
@@ -33,6 +33,23 @@ class AutoTransformers:
         .. code-block:: python
 
             autohf = AutoTransformers()
+
+            jobid_config = JobID()
+            jobid_config.set_unittest_config()
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:38]: Not intuitive. Hide this from user.

File: flaml/nlp/autotransformers.py [hunk 26]
```diff
@@ -33,6 +33,23 @@ class AutoTransformers:
...
+
+            preparedata_setting = {
+                "server_name": "tmdev",
+                "data_root_path": "data/",
+                "max_seq_length": 128,
+                "jobid_config": jobid_config,
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:44]: Hide this part from user.

File: flaml/nlp/autotransformers.py [hunk 27]
```diff
@@ -1023,7 +1022,7 @@ def fit(
         self._set_num_labels(self.task_name)
         self._fp16 = fp16
         ray.shutdown()
-        ray.init(local_mode=ray_local_mode)
+        ray.init(local_mode=ray_local_mode, num_cpus=2)
```
Comment (sonichi) [inline flaml/nlp/autotransformers.py:1025]: This constant constraint shouldn't be here. It prevents users to use more cpus or gpus.
You can move ray.shutdown and ray.init() out.

File: flaml/nlp/dataset/dataprocess_auto.py [hunk 28]
```diff
@@ -178,12 +208,14 @@ def __init__(self):
         )
 
     @classmethod
-    def from_model_and_dataset_name(cls,
-                                    data_raw,
-                                    model_checkpoint_path,
-                                    dataset_name_list: list = None,
-                                    subdataset_name=None,
-                                    **kwargs):
+    def from_model_and_dataset_name(
```
Comment (sonichi) [inline flaml/nlp/dataset/dataprocess_auto.py:211]: Does the tokenizer depend on the task type? If so, task type needs to be an input argument.

**Review by sonichi — COMMENTED** [review]
review in progress

File: flaml/nlp/result_analysis/wandb_utils.py [hunk 29]
```diff
@@ -24,80 +24,88 @@ class WandbUtils:
...
+                        group=self.wandb_group_name,
+                        name=str(WandbUtils._get_next_trial_ids()),
+                        settings=wandb.Settings(_disable_stats=True),
+                        reinit=False,
+                    )
                 else:
                     return None
             except wandb.errors.UsageError as err:
                 print(err)
                 return None
```
Comment (sonichi) [inline flaml/nlp/result_analysis/wandb_utils.py:74]: Why catch this error?

File: flaml/nlp/result_analysis/wandb_utils.py [hunk 30]
```diff
@@ -24,80 +24,88 @@ class WandbUtils:
...
+                        project=self.jobid_config.get_jobid_full_data_name(),
+                        group=os.environ["WANDB_RUN_GROUP"],
+                        settings=wandb.Settings(_disable_stats=True),
+                        reinit=False,
+                    )
                 else:
                     return None
             except wandb.errors.UsageError as err:
                 print(err)
                 return None
```
Comment (sonichi) [inline flaml/nlp/result_analysis/wandb_utils.py:107]: why catch this error?

**Review by sonichi — COMMENTED** [review]
review in progress

**Review by sonichi — COMMENTED** [review]
I skipped run_analysis.py

File: flaml/model.py [hunk 31]
```diff
@@ -273,7 +276,228 @@ def config2params(self, config: dict) -> dict:
         Returns:
             A dict that will be passed to self.estimator_class's constructor.
         """
-        return config.copy()
+        params = config.copy()
+        if "resources_per_trial" in params:
+            params.pop("resources_per_trial")
```
Comment (sonichi) [inline flaml/model.py:281]: move to TransformerEstimator's method.

File: flaml/automl.py [hunk 32]
```diff
@@ -482,6 +485,10 @@ def predict_proba(self, X_test):
         return proba
 
     def _preprocess(self, X):
+        from .data import SEQCLASSIFICATION
+        from pandas import DataFrame
+        from typing import List
```
Comment (sonichi) [inline flaml/automl.py:490]: remove

File: flaml/automl.py [hunk 33]
```diff
@@ -548,6 +557,36 @@ def _validate_data(
...
+                assert X[column].dtype.name in (
+                    "object",
+                    "string",
+                ), "If the task is an NLP task, X can only contain text columns"
+                for each_cell in X[column]:
+                    if each_cell:
+                        is_str = isinstance(each_cell, str)
+                        is_list_of_int = isinstance(each_cell, list) and sum(
+                            [isinstance(x, int) for x in each_cell]
+                        ) == len(X[column])
```
Comment (sonichi) [inline flaml/automl.py:575]:
```suggestion
                        is_list_of_int = isinstance(each_cell, list) and all(
                            isinstance(x, int) for x in each_cell)
```

File: flaml/nlp/README.md [hunk 34]
```diff
@@ -6,22 +6,22 @@ An example of using AutoTransformers:
...
-    "dataset_subdataset_name": "glue:mrpc",
-    "pretrained_model_size": "electra-base-discriminator:base",
-    "data_root_path": "data/",
-    "max_seq_length": 128,
+
+autohf_settings = {
+    "dataset_config": ["glue", "mrpc"],
+    "model_path": "google/electra-small-discriminator",
+    "output_dir": "data/",
+    "resources_per_trial": {"cpu": 1, "gpu": 1},
```
Comment (sonichi) [inline flaml/nlp/README.md:15]: n_jobs, n_gpus

**Review by sonichi — COMMENTED** [review]
I didn't review nlp/

**Review by sonichi — CHANGES_REQUESTED** [review]
Recover the files you deleted from test/

File: flaml/automl.py [hunk 35]
```diff
@@ -118,9 +120,19 @@ def update(self, result, time_used, save_model_history=False):
...
+                )
+                if n_iter is not None and "final_global_step" in config:
+                    config["final_global_step"] = n_iter
+            else:
+                n_iter = trained_estimator and trained_estimator.params.get(
+                    "n_estimators"
+                )
+                if n_iter is not None and "n_estimators" in config:
+                    config["n_estimators"] = n_iter
+
```
Comment (sonichi) [inline flaml/automl.py:135]: Use the class variable ITER_HP. Don't test the task type here.

File: flaml/automl.py [hunk 36]
```diff
@@ -1552,14 +1565,19 @@ def custom_metric(
...
             self._state.task = TS_FORECAST
         else:
             self._state.task = task
         self._state.log_training_metric = log_training_metric
+
+        from .nlp.utils import _is_nlp_task
+
+        if _is_nlp_task(task):
+            fit_kwargs["is_retrain"] = False
+
```
Comment (sonichi) [inline flaml/automl.py:1580]: Don't specialize for nlp task here. You can check the 'earlystop' branch to get a sense how to set the retrain flag.

File: flaml/data.py [hunk 37]
```diff
@@ -228,6 +228,7 @@ def fit_transform(self, X: Union[DataFrame, List[str], List[List[str]]], y, task
         from .nlp.utils import _is_nlp_task
 
         if _is_nlp_task(task):
+            X = X.copy()
```
Comment (sonichi) [inline flaml/data.py]: Don't do this here.  X can be a list.

File: flaml/model.py [hunk 38]
```diff
@@ -312,6 +314,7 @@ def search_space(cls, **params):
                 "domain": tune.loguniform(lower=1e-8, upper=1e-6),
             },
             "seed": {"domain": tune.choice(list(range(40, 45)))},
+            "final_global_step": {"domain": tune.choice([sys.maxsize])},
```
Comment (sonichi) [inline flaml/model.py:317]:
```suggestion
            "final_global_step": {"domain": sys.maxsize},

```

File: flaml/model.py [hunk 39]
```diff
@@ -334,7 +337,7 @@ def _preprocess(self, X, task, **kwargs):
 
         return tokenize_text(X, task, self.custom_hpo_args)
 
-    def _split_train_val(self, X_train, y_train, **kwargs):
+    def _get_train_val(self, X_train, y_train, **kwargs):
         """
         If the validation fold is already defined (in kwargs), set it directly
         Otherwise, split the train fold and reserve 10% dataset for validation
```
Comment (sonichi) [inline flaml/model.py]: Need update. Or, remove this function as it's straightforward.

File: flaml/model.py [hunk 40]
```diff
@@ -369,6 +371,11 @@ def on_step_end(self, args, state, control, **kwargs):
                         > self.train_begin_time + budget
                     ):
                         control.should_training_stop = True
+                        control.should_save = True
+                        control.should_evaluate = True
```
Comment (sonichi) [inline flaml/model.py:375]: remove?

File: flaml/model.py [hunk 41]
```diff
@@ -386,17 +393,22 @@ def on_step_end(self, args, state, control, **kwargs):
...
+                eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            else:
+                eval_dataset = None
         else:
             train_dataset = Dataset.from_pandas(self._join(X_train, y_train))
-            eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            if X_val is not None:
+                eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            else:
+                eval_dataset = None
```
Comment (sonichi) [inline flaml/model.py]: Why not preprocess first and then generate the train_dataset & eval_dataset? The checking of dtypes should be in _preprocess()

File: flaml/model.py [hunk 42]
```diff
@@ -470,13 +482,31 @@ def _model_init():
 
         trainer.train()
 
-        self._checkpoint_path, _ = self._get_checkpoint(trainer.ckpt_to_metric)
+        if eval_dataset is not None:
+            # if validation data is non empty, select the best checkpoint and save the final global step to self.params
+            self._checkpoint_path = self._select_checkpoint(
```
Comment (sonichi) [inline flaml/model.py:487]: why not save the last checkpoint before selecting the checkpoint?

File: flaml/nlp/huggingface/trainer.py [hunk 43]
```diff
@@ -15,23 +15,26 @@ def evaluate(self, eval_dataset=None):
...
+                    output.metrics[key[5:]] = output.metrics.pop(key)
+
+            try:
+                self.ckpt_to_metric[ckpt_dir] = output.metrics
+                self.ckpt_to_global_step[ckpt_dir] = self.state.global_step
+            except AttributeError:
+                self.ckpt_to_metric = {}
+                self.ckpt_to_metric[ckpt_dir] = output.metrics
+                self.ckpt_to_global_step = {}
+                self.ckpt_to_global_step[ckpt_dir] = self.state.global_step
```
Comment (sonichi) [inline flaml/nlp/huggingface/trainer.py:37]:
```suggestion
            if hasattr(self, "ckpt_to_global_step"):
                self.ckpt_to_metric[ckpt_dir] = output.metrics
                self.ckpt_to_global_step[ckpt_dir] = self.state.global_step
            else:
                self.ckpt_to_global_step = {ckpt_dir: self._state.global_step}
                self.ckpt_to_metric = {ckpt_dir: output.metrics}
```

**Review by sonichi — CHANGES_REQUESTED** [review]

File: flaml/automl.py [hunk 44]
```diff
@@ -1502,6 +1607,14 @@ def custom_metric(
         self._auto_augment = auto_augment
         self._min_sample_size = min_sample_size
         self._prepare_data(eval_method, split_ratio, n_splits)
+
+        if _is_nlp_task(self._state.task):
+            if X_val is not None:
+                self._state.fit_kwargs["X_val"] = X_val
+            if y_val is not None:
+                self._state.fit_kwargs["y_val"] = y_val
```
Comment (sonichi) [inline flaml/automl.py:1615]: move to _compute_with_config_base(); don't modify fit_kwargs in place

File: flaml/automl.py [hunk 45]
```diff
@@ -607,6 +689,8 @@ def _validate_data(
             self._state.groups = groups
 
     def _prepare_data(self, eval_method, split_ratio, n_splits):
+        from .nlp.utils import _is_nlp_task
```
Comment (sonichi) [inline flaml/automl.py:692]: move this import to the beginning as it is used multiple times.

File: flaml/automl.py [hunk 46]
```diff
@@ -475,7 +478,7 @@ def time_to_find_best_model(self) -> float:
         """Time taken to find best model in seconds."""
         return self.__dict__.get("_time_taken_best_iter")
 
-    def predict(self, X_test):
+    def predict(self, X_test: Union[DataFrame, List[str], List[List[str]]]):
```
Comment (sonichi) [inline flaml/automl.py:481]: we also allow numpy array

File: flaml/data.py [hunk 47]
```diff
@@ -212,7 +212,7 @@ def concat(X1, X2):
 class DataTransformer:
     """Transform input training data."""
 
-    def fit_transform(self, X: Union[DataFrame, List[str], List[List[str]]], y, task):
+    def fit_transform(self, X: DataFrame, y, task):
```
Comment (sonichi) [inline flaml/data.py:215]: X is not necessarily dataframe

File: flaml/data.py [hunk 48]
```diff
@@ -359,7 +349,7 @@ def fit_transform(self, X: Union[DataFrame, List[str], List[List[str]]], y, task
         self._task = task
         return X, y
 
-    def transform(self, X: Union[DataFrame, List[str], List[List[str]]]):
+    def transform(self, X: DataFrame):
```
Comment (sonichi) [inline flaml/data.py:352]: same problem

**Review by sonichi — CHANGES_REQUESTED** [review]

File: flaml/automl.py (hunk 44 above)
Comment (liususan091219) [reply flaml/automl.py:1644]: moved;

File: flaml/model.py (hunk 42 above)
Comment (sonichi) [reply flaml/model.py:487]: Check whether it's already checkpointed to avoid double checkpointing.

Comment (liususan091219) [reply flaml/model.py:487]: can't get the checkpoint path without saving the checkpoint using trainer.save_state()

File: flaml/model.py (hunk 41 above)
Comment (liususan091219) [reply flaml/model.py]: I removed the case where X_train.dtypes[0] != "string"

**Review by sonichi — APPROVED** [review]

**Pull request merged**